kbase-extension/static/kbase/js/widgets/kbaseTabs.js

```diff
@@ -64,167 +64,208 @@ define(['kbwidget', 'bootstrap', 'jquery', 'kbaseDeletePrompt'], (
+                return;
+            }
+            if (tab.content) {
+                $tabPanel.append(tab.content);
+                $tabPanel.hasContent = true;
+            } else if (tab.showContentCallback) {
+                try {
+                    $tabPanel.html(tab.showContentCallback($tabPanel));
+                } catch (ex) {
+                    $tabPanel.html($ErrorMessage(ex));
+                }
+                if (!tab.dynamicContent) {
+                    $tabPanel.hasContent = true;
+                }
             }
         },
 
         addTab: function (tab) {
-            if (tab.canDelete == undefined) {
+            if (typeof tab.canDelete === 'undefined') {
                 tab.canDelete = this.options.canDelete;
             }
 
-            const $tab = $('<div role="tabpanel"></div>').addClass('tab-pane fade');
-
-            $tab.hasContent = false;
+            const $tabPanel = $('<div role="tabpanel"></div>').addClass('tab-pane fade');
+
+            $tabPanel.hasContent = false;
+
+            // TODO: this is a hack, remove it!
+            // NOTE: So, I was wondering why this code is here? It is not necessary for the
+            // kbaseTabs logic. However,  there are tab usages which presume that after the tab
+            // is created, that all tab content is rendered and inserted into the DOM.
+            // E.g. in kbaseExpressionFeatureClusters the content contains links which are treated as buttons,
+            // whose event handlers are added as the tabset is rendered.
+            // Any post-tab-render, per-tab logic should be handled by the onShown or whenShown callbacks.
+            // Note that this approach will fail if a tab uses "showContentCallback".
+            // However, if we use this.showTabPanel() here to possibly invoke showContentCallback, it obviates
+            // the whole lazy nature of showContentCallback!
+            // In other words, we need to keep this here as it is, until we can fix code which relies upon
+            // this behavior.
+            // Probably better is to fork this widget and it can incrementally spread throughout the codebase
+            // until there are no usages of the old one left, then remove it.
             if (tab.content) {
-                $tab.append(tab.content);
-                $tab.hasContent = true;
+                $tabPanel.append(tab.content);
+                $tabPanel.hasContent = true;
             }
 
-            if (this.options.border) {
-                $tab.css('border', 'solid ' + this.options.borderColor);
-                $tab.css('border-width', '0px 1px 0px 1px');
-                $tab.css('padding', '3px');
+            // We need to preserve "this" for the jquery "click" handling below, since that code
+            // uses the jquery logic for "this".
+            const $widget = this;
+
+            // Create the tab button itself.
+            const $navButton = $('<a>')
```

Comment:
in order to make debugging and understanding the code construction better, I refactored out some large swathes of independent code like this.